bugfix: fix order() behavior in GlobalTransactionalInterceptorHandler to ensure correct sorting of invocation handlers

keepConcentration opened the pull request:
<!--
    Licensed to the Apache Software Foundation (ASF) under one or more
    contributor license agreements.  See the NOTICE file distributed with
    this work for additional information regarding copyright ownership.
    The ASF licenses this file to You under the Apache License, Version 2.0
    (the "License"); you may not use this file except in compliance with
    the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0
    
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
-->
<!-- Please make sure you have read and understood the contributing guidelines -->

- [x] I have registered the PR [changes](../changes).

### Ⅰ. Describe what this PR did

This PR fixes the incorrect sorting behavior of `GlobalTransactionalInterceptorHandler` in `DefaultInterfaceParser.parserInterfaceToProxy()`.
Although the `order` field was being set using `getOrder()` and `setOrder()`, it was not being reflected in the sorting logic due to the lack of an overridden `order()` method.
The fix ensures that all `ProxyInvocationHandler` implementations, including `GlobalTransactionalInterceptorHandler`, are sorted based on the intended `order` value.

### Ⅱ. Does this pull request fix one issue?
<!-- If that, add "fixes #xxx" below in the next line, for example, fixes #97. -->
fixes #7568 

This fixes the handler ordering inconsistency described in the related issue.

### Ⅲ. Why don't you add test cases (unit test/integration test)? 


### Ⅳ. Describe how to verify it


### Ⅴ. Special notes for reviews

keepConcentration committed:
bugfix: fix incorrect sorting by overriding order() in GlobalTransactionalInterceptorHandler (#7563)

YongGoose committed:
Merge branch '2.x' into feature/7568

YongGoose commented on integration-tx-api/src/main/java/org/apache/seata/integration/tx/api/interceptor/handler/AbstractProxyInvocationHandler.java:
Is there a specific reason why `order` method wasn’t implemented in `AbstractProxyInvocationHandler`?
IMO, Implementing `order` in `AbstractProxyInvocationHandler` would allow us to take advantage of inheritance.

keepConcentration replied:
I observed that both `SagaAnnotationActionInterceptorHandler` and `TccActionInterceptorHandler`, which inherit from `AbstractProxyInvocationHandler`, explicitly override the `order()` method.

To maintain consistency and clarity across all subclasses, I felt it would be more appropriate for `GlobalTransactionalInterceptorHandler` to do the same, rather than relying on a shared implementation in the abstract class.

Of course, if you believe it would be better to introduce a shared `order()` implementation in the abstract class, I'm more than happy to follow that direction.
I'm open to adjusting the PR according to your preference or the maintainers' conventions.

YongGoose replied:
I believe adding it to the abstract class would be a better approach, as this way the subclasses aren’t forced to implement it explicitly through inheritance.

YongGoose commented on integration-tx-api/src/main/java/org/apache/seata/integration/tx/api/interceptor/handler/AbstractProxyInvocationHandler.java:
```diff
@@ -59,7 +59,7 @@ public void setOrder(int order) {
     @Override
     public int getOrder() {
-        return this.order;
+        return this.order();
```
Is there a reason why the `order` function needs to return?

keepConcentration replied:
You're right — the line should return the `order` field directly instead of calling `this.order()`. I will update the code accordingly to return the `order` field as you suggested !

YongGoose reviewed (commented):
Nice work!
Left some comments.

keepConcentration commented:
Thank you for the review! Let me summarize the feedback as I understand it:
1. Implement the `order()` method in `AbstractProxyInvocationHandler`:
```java
    @Override
    public int order() {
        return this.order;
    }
```

2. Implement the `order()` method in `GlobalTransactionalInterceptorHandler`:
```java
    @Override
    public int order() {
        return this.order;
    }
```
→ With this, all classes implementing `ProxyInvocationHandler` will now explicitly implement the `order()` method.

3. The `getOrder()` method in `AbstractProxyInvocationHandler` returns `this.order`:
```java
    @Override
    public int getOrder() {
        return this.order;
    }
```
 
One thing I noticed is that both the `order()` and `getOrder()` methods in `AbstractProxyInvocationHandler` now have the same implementation, which results in a bit of duplication.

Please let me know if this approach looks good to you. I’ll proceed accordingly once I hear back from you.

YongGoose commented:
> One thing I noticed is that both the `order()` and `getOrder()` methods in `AbstractProxyInvocationHandler` now have the same implementation, which results in a bit of duplication.
> 
> Please let me know if this approach looks good to you. I’ll proceed accordingly once I hear back from you.

Since the original issue was that `getOrder` wasn’t working correctly, this approach looks fine to me.
@funky-eyes  WDYT?

YongGoose commented:
@keepConcentration 

Additionally, for the tests, it would be good to declare multiple classes inheriting from `AbstractProxyInvocationHandler`, assign different `order` values to each, and verify that they are sorted as intended.

YongGoose commented:
@keepConcentration 

When work is done, Don't hesitate to ping me :)

keepConcentration committed:
feature: implement order() method in AbstractProxyInvocationHandler (#7568)

keepConcentration committed:
test: add AbstractProxyInvocationHandler ordering test code (#7568)

keepConcentration commented:
@YongGoose

Thank you again for the detailed feedback!
I've updated the implementation based on your suggestions, including:

- Implemented the `order()` method in `AbstractProxyInvocationHandler`
- Added a test to verify sorting behavior in `AbstractProxyInvocationHandler`

Please feel free to take another look when you have time.
Let me know if there’s anything else I should adjust 🙏

YongGoose reviewed (changes requested):
Please resolve the conflicts.

keepConcentration committed:
Merge branch '2.x' into feature/7568

keepConcentration commented:
@YongGoose

Ok, I resolved!

keepConcentration committed:
test: add missing ASF license header in DefaultInterfaceParserTest

keepConcentration committed:
test: remove unused test for GlobalTransactionalInterceptorHandler order method (#7568)

keepConcentration committed:
test: add test for handler ordering and sorting logic in DefaultInterfaceParserTest (#7568)

keepConcentration committed:
Merge branch '2.x' into feature/7568

# Conflicts:
#	changes/en-us/2.x.md
#	changes/zh-cn/2.x.md

keepConcentration commented:
@YongGoose 

Thank you for the earlier feedback 🙏
I have made the following additional changes:

- Added a unit test (`testSetOrderAndSorting`) to verify sorting behavior:
  - Creates multiple `ProxyInvocationHandler` instances with different order values
  - Uses `setOrder()` to change the `order` dynamically
  - Asserts both pre-sort and post-sort ordering using the `order()` method

This should ensure that the ordering logic in `AbstractProxyInvocationHandler` works as intended.
Please review when you have time — I believe this is now ready for review ✅

YongGoose commented on integration-tx-api/src/test/java/org/apache/seata/integration/tx/api/interceptor/parser/DefaultInterfaceParserTest.java:
```diff
@@ -0,0 +1,137 @@
+import java.util.List;
+import java.util.Set;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Please avoid using `*`

keepConcentration replied:
Got it! I’ll replace the wildcard imports with explicit ones.

YongGoose reviewed (approved):
LGTM 🚀
Also, I'd appreciate it if you could revise the parts mentioned in the comments :)

keepConcentration committed:
test: specify assertEquals import in DefaultInterfaceParserTest

funky-eyes commented on integration-tx-api/src/main/java/org/apache/seata/integration/tx/api/interceptor/handler/AbstractProxyInvocationHandler.java:
```diff
@@ -28,7 +28,7 @@ public abstract class AbstractProxyInvocationHandler implements ProxyInvocationH
     protected abstract Object doInvoke(InvocationWrapper invocation) throws Throwable;
 
-    protected int order = Integer.MAX_VALUE;
+    protected int order = 0;
```
```suggestion
    protected int order = Integer.MAX_VALUE;
```
I think we should continue to use Integer.MAX_VALUE for this value, and then return 0 in the order method of GlobalTransactionalInterceptorHandler, as the order for GlobalTransactionalInterceptorHandler.

YongGoose replied:
Considering the existing handler order, this seems like a good approach.

YongGoose committed:
Update integration-tx-api/src/main/java/org/apache/seata/integration/tx/api/interceptor/handler/AbstractProxyInvocationHandler.java

Co-authored-by: funkye <364176773@qq.com>

YongGoose commented:
@funky-eyes 

The TCC transaction propagation test is failing.
Could this be related to [this comment](https://github.com/apache/incubator-seata/issues/7568#issuecomment-3170466342)?

funky-eyes commented on integration-tx-api/src/main/java/org/apache/seata/integration/tx/api/interceptor/handler/GlobalTransactionalInterceptorHandler.java:
```diff
@@ -442,4 +442,9 @@ public SeataInterceptorPosition getPosition() {
+
+    @Override
+    public int order() {
+        return this.order;
```
```suggestion
        return 0;
```

funky-eyes committed:
Update integration-tx-api/src/main/java/org/apache/seata/integration/tx/api/interceptor/handler/GlobalTransactionalInterceptorHandler.java

funky-eyes committed:
Merge branch '2.x' into feature/7568

funky-eyes commented:
> @funky-eyes
> 
> The TCC transaction propagation test is failing. Could this be related to [this comment](https://github.com/apache/incubator-seata/issues/7568#issuecomment-3170466342)?

I think the order of the test cases has changed again, which caused this issue. You can see that the exception started appearing right after this PR was merged: https://github.com/apache/incubator-seata/actions/runs/16847550947/job/47731890204

funky-eyes committed:
Merge branch '2.x' into feature/7568

funky-eyes reviewed (approved):
LGTM

funky-eyes committed:
Merge branch '2.x' into feature/7568

funky-eyes commented:
If you're using the DingTalk app, please send your DingTalk account to my email at `jianbin@apache.org`, and I will invite you to join the Seata developer group
